torch/nn/modules/container.py

```diff
@@ -187,6 +187,122 @@ def extend(self, modules):
+                x = self.choices[choice](x)
+                x = self.activations[act](x)
+                return x
+    """
+
+    def __init__(self, modules=None):
+        super(ModuleDict, self).__init__()
+        if modules is not None:
+            self.update(modules)
+
+    def __getitem__(self, key):
+        return self._modules[key]
+
+    def __setitem__(self, key, module):
+        if not isinstance(key, str):
+            raise TypeError("ModuleDict.__setitem__ key should be string, but got " +
+                            type(key).__name__)
+
+        self.add_module(key, module)
+
+    def __delitem__(self, key):
+        del self._modules[key]
+
+    def __len__(self):
+        return len(self._modules)
+
+    def __iter__(self):
+        return iter(self._modules.keys())
+
+    def __contains__(self, key):
+        return key in self._modules
+
+    def clear(self):
+        """Remove all items from the ModuleDict.
+        """
+        self._modules.clear()
+
+    def pop(self, key):
+        r"""Remove key from the ModuleDict and return its module.
+
+        Arguments:
+            key (string): key to pop from the ModuleDict
+        """
+        v = self[key]
+        del self[key]
+        return v
+
+    def keys(self):
+        r"""Return an iterable of the ModuleDict’s keys.
+        """
+        return self._modules.keys()
+
+    def items(self):
+        r"""Return an iterable of the ModuleDict’s key/value pairs.
+        """
+        return self._modules.items()
+
+    def values(self):
+        r"""Return an iterable of the ModuleDict’s values.
+        """
+        return self._modules.values()
+
+    def update(self, modules):
+        r"""Update the ModuleDict with the key/value pairs from a mapping or
+        an iterable, overwriting existing keys.
+
+        Arguments:
+            modules (iterable): a mapping (dictionary) of (string: :class:`~torch.nn.Module``) or
+                an iterable of key/value pairs of type (string, :class:`~torch.nn.Module``)
+        """
+        if not isinstance(modules, Iterable):
+            raise TypeError("ModuleDict.update should be called with an "
+                            "iterable of key/value pairs, but got " +
+                            type(modules).__name__)
+
+        if isinstance(modules, Mapping):
+            for key, module in modules.items():
+                self[key] = module
+        else:
+            for j, m in enumerate(modules):
+                if not isinstance(m, Iterable):
+                    raise TypeError("ModuleDict update sequence element "
+                                    "#" + str(j) + " should be Iterable; is" +
+                                    type(m).__name__)
+                if not len(m) == 2:
+                    raise ValueError("ModuleDict update sequence element "
+                                     "#" + str(j) + " has length " + str(len(m)) +
+                                     "; 2 is required")
```

Comment:
It would be simpler to do
```python
for k, v in modules:
  self[k] = v
```
Really, the error checking here is a bit too much I think.

Reply:
I tried to follow the python dict
```python
>>> dict([[1,2,3]])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: dictionary update sequence element #0 has length 3; 2 is required
```

Reply:
Actually, wouldn't `self._modules.update(...)` be enough for this? And it would mimic the behavior of Python dicts perfectly

Reply:
Using `self._modules.update` will skip the `add_module` and `register_parameter` validations in `module.py`.